pkg/cloud/services/ec2/instances.go

```diff
@@ -334,7 +359,7 @@ func (s *Service) findSubnet(scope *scope.MachineScope) (string, error) {
 					*subnet.SubnetId, *subnet.AvailabilityZone, *failureDomain)
 				continue
 			}
-			if scope.AWSMachine.Spec.PublicIP != nil && *scope.AWSMachine.Spec.PublicIP && !*subnet.MapPublicIpOnLaunch {
+			if wantsPublicIP && !isSubnetPublic(routeTables, *subnet.SubnetId) {
```

Comment:
@r4f4 can we use the `IsPublic` flag for the public subnet check?

```suggestion
			if wantsPublicIP && subnet.IsPublic {
```

Reply:
@mtulio not blindly. At this point all we have is an `ec2.Subnet` which doesn't have a simple `IsPublic` field. We might be able to do that if we separate the case subnetID-supplied from the subnetFilter-supplied.

Reply:
@r4f4 apologies, I should have linked the field :)

The [SubnetSpec.IsPublic](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/f67715ea3c17728fc531ec488fb0e186f31b2457/api/v1beta2/network_types.go#L413-L416) is filled when the subnet is created. Can you access in this stage the flag from the subnet ID from `subnet.SubnetId`? Something like:
```sh
s.scope.Subnets().FindByID(*subnet.SubnetId).IsPublic
```

Reply:
Aha! That makes sense. Nice suggestion, will incorporate!

Reply:
Done